### PR TITLE
Homework is ready for the cheeeeck

### PR DESCRIPTION
**Pull request by nvkaip**

### CODE REVIEW

File: hw09/LinkedList.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    @Override
+    public void add(T value, int index) {
+        if (indexIsInBounds(index)) {
+            Node<T> newNode = new Node<>(null, value, null);
+            if (isEmpty()) {
+                firstNode = newNode;
+                lastNode = newNode;
+            } else if (index == listSize) {
+                newNode.prev = lastNode;
```
Comment (Sofasmile) [inline hw09/LinkedList.java]: try to use add(T value) to simplify code

File: hw09/Node.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.homework.hw09;
+
+public class Node<E> {
```
Comment (Sofasmile) [inline hw09/Node.java]: better make this class private static nested class

File: hw09/LinkedList.java (hunk 1 above)
Comment (nvkaip) [reply hw09/LinkedList.java]: done

File: hw09/Node.java (hunk 2 above)
Comment (nvkaip) [reply hw09/Node.java]: done, but have a question for better understanding...
Is it a way of logically grouping classes that are only used in one place?

File: hw09/LinkedList.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                indexNode.prev = newNode;
+                listSize++;
+            }
+        }
+    }
+
+    @Override
+    public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
```
Comment (SevostyanovGerman) [inline hw09/LinkedList.java]: is it optimal?

File: hw09/LinkedList.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        }
+    }
+
+    @Override
+    public T get(int index) {
+        if (indexIsInBounds(index)) {
+            return nodeIteration(index).item;
+        }
+        return null;
```
Comment (SevostyanovGerman) [inline hw09/LinkedList.java]: is it ok?

File: hw09/LinkedList.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    @Override
+    public void set(T value, int index) {
+        if (indexIsInBounds(index)) {
```
Comment (SevostyanovGerman) [inline hw09/LinkedList.java]: what about null?

File: hw09/LinkedList.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                    indexNode.next.prev = indexNode.prev;
+                }
+                indexNode.prev = null;
+                indexNode.next = null;
+            }
+            indexNode.item = null;
+            listSize--;
+            return value;
+        }
+        return null;
```
Comment (SevostyanovGerman) [inline hw09/LinkedList.java]: ?

File: hw09/LinkedList.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            while (index < listSize && indexNode.item != t) {
+                indexNode = indexNode.next;
+                index++;
+            }
+            if (indexNode.item.equals(t)) {
+                System.out.println(index);
+                return remove(index);
+            }
+        }
+        return null;
```
Comment (SevostyanovGerman) [inline hw09/LinkedList.java]: return null - it's bad

Comment (nvkaip) [reply hw09/LinkedList.java]: understood. corrected

File: hw09/LinkedList.java (hunk 3 above)
Comment (nvkaip) [reply hw09/LinkedList.java]: Well, I thought that we can just take the last node from our list and assign it first node from the adding list and vice verse, but I can't understand how we can get first node from the adding list.
We don't know what implementation of List we get.
So, yes, this is optimal.

File: hw09/LinkedList.java (hunk 6 above)
Comment (nvkaip) [reply hw09/LinkedList.java]: done it more understandable, but still can't resolve problem of making it smaller

File: hw09/LinkedList.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    @Override
+    public T remove(int index) {
+        if (isEmpty() || !indexIsInBounds(index)) {
+            throw new IllegalArgumentException("Can't handle zero-size lists.");
+        }
+        Node<T> indexNode = nodeIteration(index);
+        T value = indexNode.item;
+        if (listSize > 1) {
```
Comment (nvkaip) [inline hw09/LinkedList.java]: Above is check for only one Node in the list. If there is only one Node it will not have prev or next and we will get NPE

**Pull request merged**